### PR TITLE
Support F# Records and Options

### PR DESCRIPTION
**Pull request by nordfjord**
### Adds support for F# options

1. Exposing an OptionTypeConverter for use
2. Correctly "unnesting" options in the TypeInfo class

### CODE REVIEW

File: src/HotChocolate/Core/test/Types.FSharp.Records.Test/__snapshots__/Tests.Schema can be resolved.snap [hunk 1]
```diff
@@ -0,0 +1,47 @@
+﻿schema {
+  query: Query
+}
+
+type FSharpOptionOfString {
+  get_Value: String
+  compareTo(obj: FSharpOptionOfStringInput): Int!
```
Comment (nordfjord) [inline src/HotChocolate/Core/test/Types.FSharp.Records.Test/__snapshots__/Tests.Schema can be resolved.snap:7]: This seems to come from the `IComparable` interfaces

Comment (nordfjord) [reply src/HotChocolate/Core/test/Types.FSharp.Records.Test/__snapshots__/Tests.Schema can be resolved.snap]: Added an exclusion for this 🎉

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultTypeInspector.cs [hunk 2]
```diff
@@ -27,6 +27,7 @@ public class DefaultTypeInspector : Convention, ITypeInspector
 {
     private const string _toString = "ToString";
     private const string _getHashCode = "GetHashCode";
+    private const string _compareTo = "CompareTo";
```
Comment (nordfjord) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultTypeInspector.cs:30]: This is the fix for records

File: src/HotChocolate/Core/src/Types/Internal/TypeInfo.RuntimeType.cs [hunk 3]
```diff
@@ -117,5 +117,8 @@ private static IExtendedType RemoveNonEssentialParts(IExtendedType type)
         private static bool IsOptional(IExtendedType type) =>
             type.IsGeneric &&
             typeof(Optional<>) == type.Definition;
+
+        private static bool IsOption(IExtendedType type) =>
+            type.IsGeneric && type.Definition?.Name == "FSharpOption`1";
```
[the remark this answers is not in the document]
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Internal/TypeInfo.RuntimeType.cs:122]: I need to read up on the options type in F#

**Comment by michaelstaib** [comment]
Sorry for the wait ... will have this week time for this PR.

**Comment by AlexeyRaga** [comment]
It is awesome to see the F# support happening!

**Comment by AlexeyRaga** [comment]
We are here can't wait for it to go through ;)
@michaelstaib any chance to get it reviewed/merged?

**Comment by michaelstaib** [comment]
@PascalSenn do you have time to give this one a review?

**Comment by AlexeyRaga** [comment]
@nordfjord @PascalSenn Is there anything that I can help with to make progress on this?

**Pull request closed without merging**